### PR TITLE
add a verify equivilant to prometheus visibility rules

### LINKED ISSUE #99876: bazel visibility rules need porting to make verify
See: https://github.com/kubernetes/kubernetes/pull/99561#issuecomment-791505243

xref: https://github.com/kubernetes/enhancements/issues/2420

/sig testing
/sig architecture
/area code-organization
/assign

### PR DESCRIPTION
**Pull request by BenTheElder**
#### What type of PR is this?

/kind bug
(we should have been enforcing this, the previous solution was lost in https://github.com/kubernetes/kubernetes/pull/99561)

#### What this PR does / why we need it:
Efficiently ensures that prometheus imports remain restricted per https://github.com/kubernetes/kubernetes/issues/89267 following the removal of bazel and visibility rules.

#### Which issue(s) this PR fixes:

Fixes https://github.com/kubernetes/kubernetes/issues/99876

#### Special notes for your reviewer:

We do have multiple other tools for validating imports, unfortunately none of them are suitable for this task, there's more on this in the script comments.

The short version: `verify-imports` is not capable of handling for this use case and is intended to restrict the imports _within_ a package while `verify-import-boss` can technically express this but is far too expensive to run over the entire source tree and only checks specific directories.

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/issues/2420
```

/sig testing instrumentation
/triage accepted

### CODE REVIEW

**Comment by BenTheElder** [comment]
cc @serathius xref: https://github.com/kubernetes/kubernetes/pull/99561#issuecomment-791505243 https://github.com/kubernetes/kubernetes/issues/99876#issuecomment-805582260

File: hack/verify-prometheus-imports.sh [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+  -l \
+  -Ee '"github.com/prometheus/.*"' \
+| LC_ALL=C sort -u)
+
+# check each possible package against `go list` to verify that it really is
+# an import and not an errant string
+really_failing=()
+for pkg in "${packages_probably_importing_prometheus[@]}"; do
+  if go list \
+        -f '{{ join .Imports "\n" }}{{ printf "\n" }}{{ join .TestImports "\n" }}' \
```
Comment (BenTheElder) [inline hack/verify-prometheus-imports.sh:76]: NOTE: Imports and TestImports do need to both be checked here, there is not a single field to list all imports in a package.

https://golang.org/cmd/go/#hdr-List_packages_or_modules

File: hack/verify-prometheus-imports.sh [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+        "$pkg" \
+    | grep -E \
+        '^(k8s.io/kubernetes/vendor/)?github\.com/prometheus/.*$' \
```
Comment (BenTheElder) [inline hack/verify-prometheus-imports.sh:79]: `(k8s.io/kubernetes/vendor/)?` is because:

```console
$ GO111MODULE=off go list -f '{{ join .Imports "\n" }}{{ printf "\n" }}{{ join .TestImports "\n" }}' ./cluster/images/etcd-version-monitor | grep  'github\.com/prometheus/.*'
k8s.io/kubernetes/vendor/github.com/prometheus/client_model/go

$ GO111MODULE=on go list -f '{{ join .Imports "\n" }}{{ printf "\n" }}{{ join .TestImports "\n" }}' ./cluster/images/etcd-version-monitor | grep  'github\.com/prometheus/.*'
github.com/prometheus/client_model/go
```

Currently in our scripts we have the former case, but we will need to move to the latter someday most likely.

File: hack/verify-prometheus-imports.sh [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+  ./cluster/images/etcd-version-monitor
+  ./pkg/volume/util/operationexecutor
+  ./staging/src/k8s.io/component-base/metrics
+  ./staging/src/k8s.io/component-base/metrics/legacyregistry
+  ./staging/src/k8s.io/component-base/metrics/testutil
+  ./test/e2e/apimachinery
+  ./test/e2e_node
+  ./test/integration/apiserver/flowcontrol
+  ./test/integration/metrics
+)
```
Comment (BenTheElder) [inline hack/verify-prometheus-imports.sh:48]: This is the current list of importers. 
It also matches the previous visibility rule, though is more specific for component-base/metrics.

https://github.com/kubernetes/kubernetes/blob/a1f2787897a81c375a4de28482b5eb649851800a/build/visible_to/BUILD#L431-L444

**Comment by BenTheElder** [comment]
passed in 2s in CI https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/100552/pull-kubernetes-verify/1374967748962029568

**Comment by BenTheElder** [comment]
/retest
Integration flake

**Comment by palnabarun** [comment]
@BenTheElder -- do you want this PR to be in the milestone? asking since the issue is in milestone. :)

**Comment by BenTheElder** [comment]
Yes, let me fix that.

We want to prevent regressions here after dropping bazel, the test is very cheap, and the changes are contained to `make verify` / should not be flaky.

File: hack/verify-prometheus-imports.sh (hunk 1 above)
Comment (BenTheElder) [reply hack/verify-prometheus-imports.sh:76]: Note 2: while preventing errant strings from triggering the check, it really doesn't make sense to have a string literal import like this outside of an import and we could always change the string to use backticks instead of double quotes. There are currently no cases of this.

Using go list also means we don't check imports somehow gated behind build tags, besides being mildly more complex and slower than just one pass with grep.

Comment (BenTheElder) [reply hack/verify-prometheus-imports.sh:76]: We could also consider matching individual files instead of packages which would be simpler, faster, and more restrictive. But I'm not sure if we want that.

File: hack/verify-prometheus-imports.sh [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+  # convert from file to package, and only insert unique results
+  # we want to minimize the amount of `go list` calls we need to make
+  pkg="$(dirname "${line}")"
+  if ! kube::util::array_contains "$pkg" "${packages_probably_importing_prometheus[@]}"; then
+    packages_probably_importing_prometheus+=("$pkg")
+  fi
+done < <(grep \
+  --exclude-dir={_output,vendor} \
+  --include='*.go' \
+  -R . \
```
Comment (dims) [inline hack/verify-prometheus-imports.sh:66]: replace `.` with `$KUBE_ROOT` just to be sure?

Comment (BenTheElder) [reply hack/verify-prometheus-imports.sh:66]: Good catch, thanks. Will do 👍

File: hack/verify-prometheus-imports.sh [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+# more cheaply than actually loading the entire import graph for the repo.
+packages_probably_importing_prometheus=()
+while IFS='' read -r line; do
+  # convert from file to package, and only insert unique results
+  # we want to minimize the amount of `go list` calls we need to make
+  pkg="$(dirname "${line}")"
+  if ! kube::util::array_contains "$pkg" "${packages_probably_importing_prometheus[@]}"; then
+    packages_probably_importing_prometheus+=("$pkg")
+  fi
+done < <(grep \
```
Comment (dims) [inline hack/verify-prometheus-imports.sh:63]: do all these options work well whether GNU grep or BSD grep is used?

Comment (BenTheElder) [reply hack/verify-prometheus-imports.sh]: yes. this doesn't use any GNU extensions, I tested this on mac first (2.5.1-FreeBSD) and then ~debian (GNU grep 3.6)

File: hack/verify-prometheus-imports.sh (hunk 4 above)
Comment (BenTheElder) [reply hack/verify-prometheus-imports.sh:66]: actually, I cd'ed within this subshell instead. because then we don't need to do more path mangling to get package paths instead of absolute paths.

File: hack/verify-prometheus-imports.sh [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+  if ! kube::util::array_contains "$pkg" "${packages_probably_importing_prometheus[@]}"; then
+    packages_probably_importing_prometheus+=("$pkg")
+  fi
+done < <(cd "${KUBE_ROOT}" && grep \
+  --exclude-dir={_output,vendor} \
+  --include='*.go' \
+  -R . \
+  -l \
+  -Ee '"github.com/prometheus/.*"' \
+| LC_ALL=C sort -u)
```
Comment (spiffxp) [inline hack/verify-prometheus-imports.sh:69]: I think you could xargs dirname and uniq to avoid the array logic above but not a blocker

File: hack/verify-prometheus-imports.sh (hunk 1 above)
Comment (spiffxp) [reply hack/verify-prometheus-imports.sh:76]: First review comment here could be a comment in the file for future maintainers

File: hack/verify-prometheus-imports.sh (hunk 2 above)
Comment (spiffxp) [reply hack/verify-prometheus-imports.sh:79]: This also seems like it could be a comment in file (sans invocations, just an explanation that we want to be future-proof for module support)

**Review by spiffxp — APPROVED** [review]
Non-blocking suggestions

File: hack/verify-prometheus-imports.sh (hunk 6 above)
Comment (BenTheElder) [reply hack/verify-prometheus-imports.sh:69]: `dirname` only accepts a single argument on darwin, so it's not really going to help, we still need to do it serially.

I think we're going to switch to filepaths anyhow though, after discussing what is desired with @logicalhan elsewhere.

File: hack/verify-prometheus-imports.sh (hunk 1 above)
Comment (BenTheElder) [reply hack/verify-prometheus-imports.sh:76]: this is being dropped, otherwise I agree.

Comment (BenTheElder) [reply hack/verify-prometheus-imports.sh:76]: per @logicalhan moving to allow-listing by file since we can easily do that now and do want to minimize these.

File: hack/verify-prometheus-imports.sh (hunk 2 above)
Comment (BenTheElder) [reply hack/verify-prometheus-imports.sh]: agreed, but this code is now dropped.

**Review by logicalhan — COMMENTED** [review]
/lgtm

**Comment by BenTheElder** [comment]
Changes:
- this now has a list of specific files instead of packages
- this now does not use `go list`, so we don't miss anything behind build tags
- this now fails if files in the allow-list are not actually failing anymore and should be removed from the allow-list
